include/revng/Pipeline/LLVMPipe.h

```diff
@@ -0,0 +1,178 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <array>
+#include <memory>
+#include <system_error>
+#include <utility>
+#include <vector>
+
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Analysis/ScalarEvolutionExpressions.h"
+#include "llvm/Bitcode/BitcodeReader.h"
+#include "llvm/Bitcode/BitcodeWriter.h"
+#include "llvm/IR/GlobalValue.h"
+#include "llvm/IR/LLVMContext.h"
+#include "llvm/IR/LegacyPassManager.h"
+#include "llvm/IR/Module.h"
+#include "llvm/IRReader/IRReader.h"
+#include "llvm/Linker/Linker.h"
+#include "llvm/Pass.h"
+#include "llvm/Passes/PassBuilder.h"
+#include "llvm/Support/Error.h"
+#include "llvm/Transforms/Utils/Cloning.h"
+#include "llvm/Transforms/Utils/ValueMapper.h"
+
+#include "revng/Pipeline/ContainerFactorySet.h"
+#include "revng/Pipeline/Contract.h"
+#include "revng/Pipeline/LLVMContainer.h"
+#include "revng/Pipeline/Step.h"
+#include "revng/Support/Debug.h"
+
+namespace Pipeline {
+
+class LLVMPipeBaseImpl {
+public:
+  virtual ~LLVMPipeBaseImpl() = default;
+  virtual void registerPasses(llvm::legacy::PassManager &Manager) = 0;
+  virtual const std::vector<GroupContract> &getContract() const = 0;
+  virtual std::unique_ptr<LLVMPipeBaseImpl> clone() const = 0;
+  virtual llvm::StringRef getName() const = 0;
+};
+
+/// LLVM pipes are pipes composed of any number of llvm passes.
+/// The contract is deduced by the used passes, and the passes are
+/// assembled into pass managers when needed.
+template<typename LLVMPipePass>
+class LLVMPipeImpl : public LLVMPipeBaseImpl {
+private:
+  LLVMPipePass PipePass;
+  std::vector<GroupContract> Contract;
+
+public:
+  using RegistrationFunctionType = void (*)(llvm::legacy::PassManager &);
+
+  ~LLVMPipeImpl() override = default;
+  void registerPasses(llvm::legacy::PassManager &Manager) override {
+    PipePass.registerPasses(Manager);
+  }
+
+  const std::vector<GroupContract> &getContract() const override {
+    return Contract;
+  }
+
+  std::unique_ptr<LLVMPipeBaseImpl> clone() const override {
+    return std::make_unique<LLVMPipeImpl>(*this);
+  }
+
+  LLVMPipeImpl(LLVMPipePass Pass) :
+    PipePass(std::move(Pass)), Contract(this->PipePass.getContract()) {}
+
+  LLVMPipeImpl(const LLVMPipeImpl &Other) = default;
+  LLVMPipeImpl(LLVMPipeImpl &&Other) = default;
+
+  template<typename... ArgsT>
+  LLVMPipeImpl(ArgsT &&...Args) :
+    PipePass(std::forward<ArgsT>(Args)...),
+    Contract(this->PipePass.getContract()) {}
+
+  llvm::StringRef getName() const override { return LLVMPipePass::Name; }
+};
+
+/// Implementation of the llvm pipes to be instantiated for a particular llvm
+/// container
+template<typename LLVMContainer>
+class LLVMPipeBase {
```

Comment:
i was trying to write a concept with pipes, the issue is that the concept must handle the variadic run mathod.

i can write a template that returns  a bool to check if the run method is correct, but then the error message will be "the run method does not respect the condition X" and then the condition X is some mysterious template that is checking each argument to see if it is a container or not. 
It *should* be possible to write recursive concepts so that they can iterate on function arguments, not sure if we want to go there toh.